### PR TITLE
LTO CancelCrossCUDebugInfo

### PR DESCRIPTION
**Pull request by BgZun**
Additon of LTO CancelCrossCUDebugInfo pass

### CODE REVIEW

File: llvm/lib/Transforms/Utils/LTOCuPass.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+// End copied
+
+PreservedAnalyses CancelCrossCUDebugInfo::run(Module &M,
+                                              ModuleAnalysisManager &MA) {
+
+  M.print(errs(), nullptr);
+  for (Function &F : M) {
+    if (isFunctionSkipped(F))
```
Comment (cme) [inline llvm/lib/Transforms/Utils/LTOCuPass.cpp]: This returns from processing the whole module if there's a single function in the module that's skipped. Possibly this was left in from a change that previously was doing this per-function?

File: llvm/lib/Passes/PassBuilder.cpp [hunk 2]
```diff
@@ -1649,6 +1655,7 @@ PassBuilder::buildLTODefaultPipeline(OptimizationLevel Level,
                                      ModuleSummaryIndex *ExportSummary) {
   ModulePassManager MPM;
 
+  errs()<<"KDLAKMJKDAJOFGJAOIGJMAOGMJAHAOHAO"<<'\n';
```
Comment (cme) [inline llvm/lib/Passes/PassBuilder.cpp]: Spurious debug output

File: llvm/include/llvm/InitializePasses.h [hunk 3]
```diff
@@ -455,7 +456,6 @@ void initializeWinEHPreparePass(PassRegistry&);
 void initializeWriteBitcodePassPass(PassRegistry&);
 void initializeWriteThinLTOBitcodePass(PassRegistry&);
 void initializeXRayInstrumentationPass(PassRegistry&);
-
```
Comment (djtodoro) [inline llvm/include/llvm/InitializePasses.h]: remove this from the patch (no need to add/remove this empty lines)

File: llvm/lib/LTO/LTOBackend.cpp [hunk 4]
```diff
@@ -82,7 +83,12 @@ static cl::opt<std::string> LtoExternalAsm("lto-external-asm",
 static cl::list<std::string> LtoExternalAsmArgs("lto-external-asm-arg",
                                                 cl::desc("Args to pass to "
                                                          "external assembler in LTO backend"));
-  
+                                                    
+static cl::opt<bool>
+    EnableCancelCrossCUDebugInfo("enable-cancel-cross-cu-debug-info",cl::init(false), cl::Hidden, cl::ZeroOrMore,
+                                 cl::desc("Enable cross cu canceling"));
```
Comment (djtodoro) [inline llvm/lib/LTO/LTOBackend.cpp]: please run clang-format

File: llvm/lib/Passes/PassBuilder.cpp [hunk 5]
```diff
@@ -1717,8 +1718,8 @@ PassBuilder::buildLTODefaultPipeline(OptimizationLevel Level,
   }
 
   // Now deduce any function attributes based in the current code.
-  MPM.addPass(createModuleToPostOrderCGSCCPassAdaptor(
-              PostOrderFunctionAttrsPass()));
+  MPM.addPass(
+      createModuleToPostOrderCGSCCPassAdaptor(PostOrderFunctionAttrsPass()));
```
Comment (djtodoro) [inline llvm/lib/Passes/PassBuilder.cpp]: why this refactor?

File: llvm/lib/Passes/PassBuilder.cpp [hunk 6]
```diff
@@ -2195,7 +2196,8 @@ Expected<SimplifyCFGOptions> parseSimplifyCFGOptions(StringRef Params) {
         return make_error<StringError>(
             formatv("invalid argument to SimplifyCFG pass bonus-threshold "
                     "parameter: '{0}' ",
-                    ParamName).str(),
+                    ParamName)
+                .str(),
```
Comment (djtodoro) [inline llvm/lib/Passes/PassBuilder.cpp]: why this as well

File: llvm/lib/Passes/PassRegistry.def [hunk 7]
```diff
@@ -46,6 +46,7 @@ MODULE_PASS("attributor", AttributorPass())
 MODULE_PASS("annotation2metadata", Annotation2MetadataPass())
 MODULE_PASS("openmp-opt", OpenMPOptPass())
 MODULE_PASS("called-value-propagation", CalledValuePropagationPass())
+MODULE_PASS("ccudi", CancelCrossCUDebugInfo())
```
Comment (djtodoro) [inline llvm/lib/Passes/PassRegistry.def]: please add reasonable name to the option, something like `cancel-cross-cu-debug-info`

File: llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+#include "llvm/IR/DebugInfoMetadata.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+#include "llvm/Transforms/Utils.h"
+#include "llvm/Transforms/Utils/ValueMapper.h"
+
+using namespace llvm;
+
+// Copied from debugify
```
Comment (djtodoro) [inline llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp]: please remove this

File: llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+bool isFunctionSkipped(Function &F) {
+  return F.isDeclaration() || !F.hasExactDefinition();
+}
+
+// End copied
```
Comment (djtodoro) [inline llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp]: this as well

File: llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp [hunk 10]
```diff
@@ -0,0 +1,139 @@
+//===- CancelCrossCUDebugInfo.cpp - Enable cross cu canceling
+//--------------------===//
```
Comment (djtodoro) [inline llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp]: fix this

File: llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    auto mapMDtoMD = [&VMap](MDNode *Old, MDNode *New) {
+      // Avoid clobbering an existing mapping.
+      (void)VMap.MD().try_emplace(Old, New);
+    };
+    DISubprogram *SPInline = nullptr;
+    for (BasicBlock &BB : F) {
+      for (Instruction &I : BB) {
+        auto DebugLoc = I.getDebugLoc();
+
+        // Inline check
```
Comment (djtodoro) [inline llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp]: remove this

File: llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    if (isFunctionSkipped(F))
+      continue;
+
+    Module &M = *F.getParent();
+    DICompileUnit *cu = *M.debug_compile_units_begin();
+
+    DIBuilder DIB(M, true, cu);
+    LLVMContext &Ctx = M.getContext();
+
+    std::map<StringRef, DISubprogram *> FunctionNameMap;
```
Comment (cme) [inline llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp]: This seems to be used to make *filename* of the compile unit to a single DISubprogram? It seems to support only a single function per source file?

File: llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+              if (FunctionNameMap.find(T->getFilename()) ==
+                  FunctionNameMap.end()) {
+
+                SPInline = DIB.createFunction(
+                    OrigSP->getFile(), T->getName(), T->getName(),
+                    OrigSP->getFile(), T->getLine(), T->getType(),
+                    T->getScopeLine(), DINode::FlagZero, T->getSPFlags());
+
+                for (DINode *DN : T->getRetainedNodes()) {
```
Comment (cme) [inline llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp]: Are 'retained nodes' *necessarily* parameters? I can't find any reference for this intention or indeed the function in the documentation, and the name doesn't match the potential intention.

**Review by cme — COMMENTED** [review]
A question on design. We've seen that cross-CU references which cause GDB to choke include inlined subprograms, and also global variables and types. The approach here is specific to inlined subprograms, and will leave shared DI for globals and types.

A more generalised walk of the DI metadata with a map of  `pair<DICompileUnit*, DINode*>` -> `DINode*`  (starting from each entry point and creating a unique copy of `DINode` if not already available in `make_pair(ThisCU,Node)`) I thought would be simpler and more rigorous, without needing to know anything about the structure of  the DI?

File: llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+        if (InlineLoc) {
+          auto Scope = DebugLoc.getScope();
+          if (auto *T = dyn_cast<DISubprogram>(Scope)) {
+            if (T->getFilename() != OrigSP->getFilename()) {
+
+              if (FunctionNameMap.find(T->getFilename()) ==
+                  FunctionNameMap.end()) {
+
+                SPInline = DIB.createFunction(
+                    OrigSP->getFile(), T->getName(), T->getName(),
```
Comment (vdelvecc) [inline llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp]: AFAICT, this is not the containing DICompileUnit; that is in the DIB.  It is wrong to change the filename (either scope or file) and will break debugging.  We only want to replicate any content (including the DIFile) into the other CompileUnit.

File: llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+
+PreservedAnalyses CancelCrossCUDebugInfo::run(Module &M,
+                                              ModuleAnalysisManager &MA) {
+
+  for (Function &F : M) {
+    if (isFunctionSkipped(F))
+      continue;
+
+    Module &M = *F.getParent();
+    DICompileUnit *cu = *M.debug_compile_units_begin();
```
Comment (vdelvecc) [inline llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp]: Presumably in an LTO compilation there can be multiple CompileUnits.  I don't think we always want the first?

File: llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+      (void)VMap.MD().try_emplace(Old, New);
+    };
+    DISubprogram *SPInline = nullptr;
+    for (BasicBlock &BB : F) {
+      for (Instruction &I : BB) {
+        auto DebugLoc = I.getDebugLoc();
+
+        auto InlineLoc = DebugLoc.getInlinedAt();
+
+        if (InlineLoc) {
```
Comment (vdelvecc) [inline llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp:111]: There can be multiple levels of inlining recorded.  Presumably others might also be wrong?

File: llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+          auto Scope = DebugLoc.getScope();
+          if (auto *T = dyn_cast<DISubprogram>(Scope)) {
+            if (T->getFilename() != OrigSP->getFilename()) {
+
+              if (FunctionNameMap.find(T->getFilename()) ==
+                  FunctionNameMap.end()) {
+
+                SPInline = DIB.createFunction(
+                    OrigSP->getFile(), T->getName(), T->getName(),
+                    OrigSP->getFile(), T->getLine(), T->getType(),
```
Comment (vdelvecc) [inline llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp]: Types are not being replicated here.  Is this assuming the existing Type reduplication fix is being applied?

File: llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp (hunk 13 above)
Comment (djtodoro) [reply llvm/lib/Transforms/Utils/CancelCrossCUDebugInfo.cpp]: I think no. There can be `divariables` and `dilabels`.

**Comment by djtodoro** [comment]
Abandon this in favor of the `AsmPrinter`/`DwarfDebug` patch.

**Pull request closed without merging**